### ISSUE TITLE
Trean outpath `-` as stdout and inpath `-` as stdin

### ISSUE THREAD

**Issue by flying-sheep** [bskinn/sphobjinv#131]
Would be cool if specifying `-` there would not create a file called `-` but write to stdout.

One concern: You currently log to stdout instead of stderr. Therefore currently doing `sphobjinv convert json ... /dev/stdout` results in the json being mixed with log messages. This can be fixed in two ways:

1. Do all logging on stderr (I like doing that anyway in my CLI stuff)
2. make `-` as outpath imply `-q`

**Comment by bskinn** [comment]
Indeed, I'd had the same though a while back (#74), but hadn't needed the functionality myself. Now that I know it's not YAGNI, I'll prioritize it.

I like the plan of switching logging output over to `stderr`; I likely should have been doing it that way from the outset.

I have a number of things that I want to roll into a v2.1 release; I'll make this feature one of them.


Thanks for the feature request!

**Comment by bskinn** [comment]
Imported from #74:

`-` arguments for input and/or output, maybe.

`sphobjinv convert json -`  for `stdin` and `stdout` both

`sphobjinv convert zlib objects_python.inv -` for file read, `stdout` output

`sphobjinv convert plain - objects.txt` for `stdin` read, file output.

**Comment by flying-sheep** [comment]
Sounds great! Any reason you do your own CLI parsing instead of using something abstracting it like click?

With it, we could just use [File](https://click.palletsprojects.com/en/7.x/arguments/#file-args) parameters:

```py
import click

@click.group()
...
def cli(common, args):
   ...

@cli.group()
@click.argument('mode', type=click.Choice(['zlib', 'plain', 'json']))
@click.argument('in_path', type=click.File('rb'))
@click.argument('out_path', type=click.File('wb'))
@click.option(...)
...
def convert(mode, in_path, out_path, ...):
    ...
```

**Comment by bskinn** [comment]
I think I went with `argparse` for two main reasons... I came across it first, and I was also trying to minimize external dependencies where I could. Since `sphobjinv` is a library as well as a CLI tool, I was (and still am) inclined to avoid CLI-specific dependencies.

Hard to argue with the convenience & terseness, though, and I have been wanting to get familiar with `click`.  I'll think about a refactor there.

**Comment by bskinn** [comment]
Hm. Looking at the `OUTFILE` handling, I'm not sure if `click.File` would actually work. I have the CLI implemented such that a user can just pass a directory path, in which case a default output filename is constructed. Can `File` (or some other `click` feature) do that?

**Comment by flying-sheep** [comment]
Ah, in that case you’d need to handle it yourself, either by subclassing `click.File` and [overriding `convert`](https://github.com/pallets/click/blob/297150c7566a38fb672828a36b681792513d99b0/src/click/types.py#L591-L621) or by providing `callback` instead of `type` [as seen here](https://click.palletsprojects.com/en/7.x/api/#click.confirmation_option).

But yeah, if you don’t want to use it, you need to refactor `write_json` and friends to either handle `-` individually or to take open file objects and make sure the file is closed in the end.

**Comment by bskinn** [comment]
Initial POC is implemented on `add-stdio` branch. Could you give it a spin, see if it's behaving how you want/need? I haven't dealt yet with the logging to stdout vs stderr; that'll come.

**Comment by bskinn** [comment]
Same POC is now on PyPI as v2.1a1; can install with `pip install --pre`.

**Comment by flying-sheep** [comment]
Hi! So some feedback:

- it still writes to stderr when telling it to emit JSON to stdout:

  ```console
  $ PYTHONPATH=src python -m sphobjinv convert json -u https://rpy2.github.io/doc/latest/html/ - | head -c 200

  No inventory at provided URL.
  Attempting "https://rpy2.github.io/doc/latest/html/objects.inv" ...
  Remote inventory found.
 
  {"project": "rpy2", "version": "3.3.1", "count": 294, "0": {"name": "rpy2.Exception ignored in: <_io.TextIOWrapper name='<stdout>' mode='w' encoding='utf-8'>
  ```

- Why duplicate the writing? at some point all three `write_*` functions open a file and write to it. They could all just take an open file instead, somthing like:

  ```py
  if params[MODE] in binary_modes:
      file = sys.stdout.buffer if path == '-' else open(path, 'wb')
  else:
      file = sys.stdout if path == '-' else open(path, 'wt')
  try:
      if mode == ZLIB:
          write_zlib(inv, file, ...)
      ...
  except BrokenPipeError:
      pass
  except Exception:
      selective_print(...)
      sys.exit(1)
  finally:
      file.close()
  ```

**Comment by bskinn** [comment]
> *except for write_json, but write_json should open a file in binary mode, **as JSON is always utf-8**.

This is really good to know. I've been mentally wrestling with encodings a lot. I'm thankful that the Sphinx team used UTF-8 for the inventory files... I'm pretty sure it's made things waaaaay simpler.

**Comment by bskinn** [comment]
Feature implemented a/o 97c9917 (v2.1a2)

Tests to be added are being tracked in #137.

**Comment by flying-sheep** [comment]
Very cool, thank you!

**Comment by bskinn** [comment]
FYI, this feature is now available in the non-prerelease v2.1 up on PyPI.
